### PR TITLE
[WIP] Various tweaks to command argument parsing and event flow

### PR DESCRIPTION
**Pull request by zml2008**
This PR covers a few long-standing issues about command argument format and hopefully command argument handling.

The ones I want to get are:
- [x] #604 
- [x] #583 
- [ ] Adding more tags to the various location-related command arguments
- [x] #689 -- An event that is called for chat as well as commands (very preprocessey) -- completed, this is covered by #957
- [ ] #780 
- [x] ~~Only parse escapes for characters that would otherwise have a special function~~
- [ ] Refactor CommandArgs + CommandContext to be less dumb (see [PEX wip on this](https://github.com/PEXPlugins/PermissionsEx/tree/feature/elementresult) (though what happens from this will be different))
- [ ] #924 
- [x] #989 

notable todos:
- Implement `#spawn:<world>` on LocationCommandElement
- Properly handle tabcompletion on LocationCommandElement

### CODE REVIEW

**Comment by JBYoshi** [comment]
It doesn't  look like `GenericArguments.user()` supports users that haven't connected yet. Is that intended?

**Comment by zml2008** [comment]
yeah, that's intented @JBYoshi -- we would need to be passing around Futures in command parsing which gets kinda complicated and makes matching harder.

**Comment by Bammerbom** [comment]
Can a way to get the command label used be added?
Would make my life a lot easier.

File: src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java [hunk 1]
```diff
@@ -31,16 +31,26 @@
 import com.google.common.base.Function;
 import com.google.common.base.Joiner;
 import com.google.common.base.Optional;
+import com.google.common.base.Predicate;
```
Comment (Phroa) [inline src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java]: should migrate to Java's

File: src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java [hunk 2]
```diff
@@ -108,6 +118,17 @@ public static CommandElement player(Text key, Game game) {
     }
 
     /**
+     * Expect an argument to represent a player who has been online at some point, as a {@link User}
```
Comment (Phroa) [inline src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java]: Period at the end

File: src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java [hunk 3]
```diff
@@ -108,6 +118,17 @@ public static CommandElement player(Text key, Game game) {
...
+     *
+     * @param key The key to store under
+     * @param game The game to find users in
+     * @return the argument
```
Comment (Phroa) [inline src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java]: Capital `t`

**Comment by simon816** [comment]
@zml2008 This has a fair few features that would be useful, can we get an update?

**Comment by kashike** [comment]
Perhaps to be semi-consistent in command arguments, provide `Text` methods in `CommandContext` that just call their `String` counterparts (since `GenericArguments` methods take `Text` for i18n)?

**Comment by zml2008** [comment]
Another thing that came up: Possibly make CommandContext and CommandArgs interfaces (or whatever happens after their possible merger) for better interop with things like intake

**Comment by ryantheleach** [comment]
Is it currently possible to create your own CommandCallable, using the sponge CommandSpec system as a delegate without side effects like errors being sent to the user? It might simplify creation of CommandCallables for people who only need a slight tweak to how the sponge system works.

**Comment by simon816** [comment]
@zml2008 2 of 9 tasks complete. Need this (and related issues) to be completed in the next 2 days.

**Comment by zml2008** [comment]
The checklist is kinda a bad representation for progress -- there's one big thing then 8 little things.

**Comment by gabizou** [comment]
@zml2008 Just to note, when merging this, this will have to be merged into [bleeding](https://github.com/SpongePowered/SpongeAPI/tree/bleeding) for the next release.

**Comment by zml2008** [comment]
Despite GitHub disagreeing, this PR is now based on the bleeding branch.

**Comment by gabizou** [comment]
@zml2008 good to know, how close is it to being ready for merging?

**Comment by zml2008** [comment]
Getting closer -- I'm working on updating common's commands to deal with these changes and getting into some rough areas that need to be figured out rn.

**Comment by Aaron1011** [comment]
what

**Comment by zml2008** [comment]
This PR is superseded by #1116 which has a more useful diff.

**Pull request closed without merging**